Bug 1880410: convert pipeline visualization node to SVG

rottencandy opened the pull request:
**Fixes**: 
https://issues.redhat.com/browse/ODC-3661
<!-- For e.g Fixes: https://issues.redhat.com/browse/ODC-XXX -->

**Analysis / Root cause**: 
There are several rendering issues with the Pipeline Visualization Nodes (aka Tasks). This needs to be handled better and the whole thing is currently rendered through a foriegnObject inside the SVG visualization.
<!-- Briefly describe analysis of US/Task or root cause of Defect -->

**Solution Description**: 
Convert pipeline visualization node to SVG.
<!-- Describe your code changes in detail and explain the solution -->

**Screen shots / Gifs for design review**: 
(Should ideally be unchanged)
![tmp](https://user-images.githubusercontent.com/20013884/93600586-830d5880-f9dd-11ea-9591-b8565cd5589c.png)
![tmp0](https://user-images.githubusercontent.com/20013884/95361627-992a7c80-08ea-11eb-893c-7acf60ba2e5d.png)
![tmp](https://user-images.githubusercontent.com/20013884/95884763-1eee7200-0d9a-11eb-9029-c375673b67ec.png)
![tmp2](https://user-images.githubusercontent.com/20013884/95361657-a0518a80-08ea-11eb-9f5e-7d10fc280c32.png)

<!-- If change affects UI in any way, tag @openshift/team-devconsole-ux and add screenshots/gifs  -->

**Unit test coverage report**: 
Unchanged.
<!-- Attach test coverage report -->

**Browser conformance**: 
<!-- To mark tested browsers, use [x] -->
- [x] Chrome
- [x] Firefox
- [ ] Safari
- [ ] Edge

rottencandy commented:
/hold
Will be worked on in the next sprint.

rottencandy commented:
/bugzilla refresh

rottencandy commented:
/hold cancel

rottencandy commented:
/kind bug

rottencandy commented:
/retest

jerolimov commented:
Code change looks valid, but I find some UI difference.

**In view mode:**
Before:
![pr-6676-view-before](https://user-images.githubusercontent.com/139310/95621301-bfc6ef80-0a71-11eb-842e-c58885e12aab.png)
Now:
![pr-6676-view-after](https://user-images.githubusercontent.com/139310/95621311-c2c1e000-0a71-11eb-9c3e-85909111cea3.png)

1. The vertical alignment is not centered anymore, can you center it by increasing the (line) height or the y-offset here?
2. The old version uses the complete width of the boxes for longer task names (first box, I repeated the task name multiple times).
3. There is a little bit "more shadow" in the old version which is shown in the new one only on hover. I like it! I would prefer if we accept the new version.

**In edit mode:**
Before:
![pr-6676-edit-before](https://user-images.githubusercontent.com/139310/95621318-c5243a00-0a71-11eb-95af-3bf93d1608d9.png)
Now:
![pr-6676-edit-after](https://user-images.githubusercontent.com/139310/95621327-c9e8ee00-0a71-11eb-972a-effe7023a0c2.png)

1. I think the vertical alignment could also be one pixel (sorry :P) lower.
2. I don't know why, but the "really long" task name overlaps the box. Only in the edit case when it does not find the task reference!
3. The bottom border is now bigger than in the old version also when the box is not selected. Not bad, but different.

rottencandy commented:
Thanks for the review @jerolimov !
>1. The vertical alignment is not centered anymore, can you center it by increasing the (line) height or the y-offset here?

Updated.
![tmp](https://user-images.githubusercontent.com/20013884/95780618-dbd4c600-0ce9-11eb-8b28-f734c5b54af8.png)
 
>2. The old version uses the complete width of the boxes for longer task names (first box, I repeated the task name multiple times).

Updated the max length. I think there will be some differences here. The old version used a CSS solution(using `text-overflow` considers font width) but the new one uses a JS solution(does not consider font width), there seems to be no way to truncate text in SVG.

>3. There is a little bit "more shadow" in the old version which is shown in the new one only on hover. I like it! I would prefer if we accept the new version.

I tried to replicate the old version which also had hover shadow, albeit was very subtle. I can add the non-hover shadow as well if you think it's better.

> **In edit mode:**
>1. I think the vertical alignment could also be one pixel (sorry :P) lower.

Updated.

>2. I don't know why, but the "really long" task name overlaps the box. Only in the edit case when it does not find the task reference!

Thank you for catching this, I seem to have missed it. Updated.
![tmp0](https://user-images.githubusercontent.com/20013884/95781655-ed1ed200-0ceb-11eb-95c0-9e7947e1ef1c.png)

>3. The bottom border is now bigger than in the old version also when the box is not selected. Not bad, but different.
Not sure why it appears bigger, I've set it to the smallest possible width of `1px`(under `__trigger-underline` in `TaskListNode.scss`). I'll go through SVG attrs again to see if I find something.

rottencandy commented:
Also about the bottom border, I think there is a bit of inconsistency between browsers.
Didn't notice this earlier since I was testing mostly on Firefox, the button appears slightly different on Chromium.

Firefox vs. Chromium:
![inconsistency](https://user-images.githubusercontent.com/20013884/95781989-7e8e4400-0cec-11eb-8723-6da9c784f170.png)

Should I follow the Firefox appearance, or the Chromium one?
cc: @andrewballantyne

rottencandy commented:
/retest

jerolimov reviewed (approved):
Updated version looks fine for me. :+1: 

/lgtm

rottencandy commented:
/assign @andrewballantyne

rottencandy commented:
Rebased.

jerolimov reviewed (approved):
Tested it again locally, works fine.
/lgtm

jerolimov commented:
/retest

andrewballantyne commented:
@rottencandy The pipeline-plugin exists now, you'll need to rebase your code over there. Should be a straightforward endeavour as the folder structure should be relatively unchanged.

andrewballantyne commented:
/retitle [WIP] Bug 1880410: convert pipeline visualization node to SVG
/hold

Discussion with Saud has lead me to putting a WIP on this PR. There are some technical issues (beyond the rebase on the new pipeline-plugin package). Currently not at the top of our priority list. Putting on the back burner to see if we can get some time near the end of this sprint.

andrewballantyne commented on frontend/packages/pipelines-plugin/src/components/pipelines/detail-page-tabs/pipeline-details/StatusIcon.tsx:
```diff
@@ -11,17 +12,25 @@ import {
+      return (
+        <SyncAltIcon
+          {...props}
+          className={cx('fa-spin', { 'pipeline-status-icon__spin': shiftOrigin })}
```
```suggestion
          className={cx('fa-spin', { 'pipeline-status-icon--spin': shiftOrigin })}
```

BEM is 
 - Block (`pipeline-status-icon`)
 - Element (`pipeline-status-icon__someWrapper`) -- kinda like a sub Block
 - Modifier (`pipeline-status-icon--spin`) -- when you want it to be "modified" based on a prop

rottencandy replied:
Updated.

andrewballantyne reviewed (changes requested):
Having some hovering issues.

![image](https://user-images.githubusercontent.com/8126518/105774315-2396fa80-5f33-11eb-9fdc-a7b989383836.png)

When I hover over an item and add a parallel item, and then mouse up over the node and slowly to the dropdown, it removes and reappears the "+" icons on the above node.

andrewballantyne commented:
Ah, it's the tooltip capturing the mouse-over and I think it persists at that point. 

![Uploading tooltipCapture.gif…]()

andrewballantyne commented:
/hold cancel

andrewballantyne commented:
FWIW, frontend test failure is lack of `yarn i18n`

rottencandy commented:
> Ah, it's the tooltip capturing the mouse-over and I think it persists at that point.

Replacing the `onMouse` handlers with `useHover` seems to fix it.

andrewballantyne commented on frontend/packages/pipelines-plugin/src/components/pipelines/pipeline-topology/TaskListNode.scss:
```diff
@@ -5,18 +5,26 @@
+  }
+  &__trigger-underline {
+    fill: var(--pf-global--Color--100);
+    &__hover {
```
```suggestion
    &--hover {
```

rottencandy replied:
Updated.

andrewballantyne reviewed (changes requested):
Tiny nit, will test it out once I finish compiling 🙂

andrewballantyne commented:
/cc @karthikjeeyar @jerolimov 

Please one of you toss a review on this as well. If I don't see any review by EOD tomorrow, I'll LGTM it.

jerolimov commented:
/retest

karthikjeeyar commented:
Verified locally, works fine. Thanks @rottencandy 
/lgtm
